NumLib/Fem/FiniteElement/TemplateIsoparametric.h

```diff
@@ -101,13 +103,40 @@ class TemplateIsoparametric
      */
     template <ShapeMatrixType T_SHAPE_MATRIX_TYPE>
     void computeShapeFunctions(const double* natural_pt, ShapeMatrices& shape,
-                               const unsigned global_dim) const
+                               const unsigned global_dim,
+                               bool is_axially_symmetric) const
     {
         NaturalCoordsMappingType::template computeShapeMatrices<
             T_SHAPE_MATRIX_TYPE>(*_ele, natural_pt, shape, global_dim);
+        computeIntegralMeasure(is_axially_symmetric, shape);
+    }
+
+    double interpolateZerothCoordinate(
+        typename ShapeMatrices::ShapeType const& N) const
+    {
+        auto* nodes = _ele->getNodes();
+        typename ShapeMatrices::ShapeType rs(N.size());
+        for (int i=0; i<rs.size(); ++i) {
+            rs[i] = (*nodes[i])[0];
+        }
+        auto const r = N.dot(rs);
+        return r;
```

Comment:
need
```
if (r == 0.)
  r = 0.5/pi
```
for line integration at r = 0, which is used to apply a distributed source term along the axis.

In addition the case of integration point at edges (only exist in triangle) must be avoid, which are secured with the present  integration point definitions in GaussLegendreTri.cpp.